lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -0,0 +1,211 @@
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponse;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponses;
+import com.palantir.atlasdb.timelock.api.TimestampLeasesRequest;
+import com.palantir.atlasdb.timelock.api.TimestampLeasesResponse;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.client.InternalMultiClientConjureTimelockService;
+import com.palantir.lock.client.LeasedLockToken;
+import com.palantir.lock.client.TimeLockUnlocker;
+import com.palantir.lock.v2.ImmutableTimestampLeaseResult;
+import com.palantir.lock.v2.ImmutableTimestampLeaseResults;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimestampLeaseResult;
+import com.palantir.lock.v2.TimestampLeaseResults;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.ids.UniqueIds;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import org.immutables.value.Value;
+
+final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer {
+    private static final SafeLogger log = SafeLoggerFactory.get(TimestampLeaseAcquirerImpl.class);
+
+    private final Retryer<RequestAndResponse> retryer = RetryerBuilder.<RequestAndResponse>newBuilder()
+            .retryIfResult(requestAndResponse -> !wasRequestFullyFulfilled(requestAndResponse))
+            .retryIfException()
+            .withStopStrategy(StopStrategies.stopAfterAttempt(3))
+            .build();
+
+    private final Namespace namespace;
+    private final InternalMultiClientConjureTimelockService service;
+    private final TimeLockUnlocker unlocker;
+
+    TimestampLeaseAcquirerImpl(
+            Namespace namespace, InternalMultiClientConjureTimelockService service, TimeLockUnlocker unlocker) {
+        this.namespace = namespace;
+        this.service = service;
+        this.unlocker = unlocker;
+    }
+
+    @Override
+    public TimestampLeaseResults acquireNamedTimestampLeases(Map<TimestampLeaseName, Integer> requests) {
+        TimestampLeaseResponses responses = acquireNamedTimestampLeasesWithRetry(requests);
+
+        LockToken lock = createFromLeaseGuarantee(responses.getLeaseGuarantee());
+
+        Map<TimestampLeaseName, TimestampLeaseResult> results = KeyedStream.stream(
+                        responses.getTimestampLeaseResponses())
+                .map((timestampName, response) -> createTimestampLeaseResult(requests.get(timestampName), response))
+                .collectToMap();
+
+        return ImmutableTimestampLeaseResults.of(lock, results);
+    }
+
+    @Override
+    public void close() {
+        unlocker.close();
```

Comment:
This does not feel appropriate, you are borrowing the unlocker, you don't own it.

Reply:
See my published comments. I think we want our own unlocker here possibly